truffle/src/main/java/org/jruby/truffle/core/regexp/RegexpNodes.java

```diff
@@ -426,7 +430,25 @@ public Object matchGeneric(VirtualFrame frame, DynamicObject regexp, DynamicObje
                 toStrNode = insert(ToStrNodeGen.create(getContext(), getSourceSection(), null));
             }
 
-            return match(regexp, toStrNode.executeToStr(frame, other));
+            return matchWithStringCopy(regexp, toStrNode.executeToStr(frame, other));
+        }
+
+        // Creating a MatchData will store a copy of the source string. It's tempting to use a rope here, but a bit
+        // inconvenient because we can't work with ropes directly in Ruby (useful for implementing some MatchData
+        // methods). Likewise, we need to taint objects based on the source string's taint state. We mustn't allow the
+        // source string's contents to change, however, so we must ensure that we have a private copy of that string.
+        private Object matchWithStringCopy(DynamicObject regexp, DynamicObject string) {
+            final Matcher matcher = createMatcher(getContext(), regexp, string);
+            final int range = StringOperations.rope(string).byteLength();
+
+            final DynamicObject matchData = matchCommon(getContext(), this, makeSubstringNode, regexp, string, true, matcher, 0, range);
+            setLastMatchNode.executeSetLastMatch(matchData);
+
+            if (matchData != nil()) {
+                return matcher.getBegin();
+            }
```

Comment:
The comment already says we need to make a defensive copy to prevent modifications. I take it you're looking for more details about that?

Re: the substring.... Unfortunately, it's not that easy because that substring method is really only dealing with things at the rope level and won't carry forward the taint state. We can deal with that within matchCommon, but I don't think it'll wind up much cleaner and passing a frame around gets complicated with the boundary in place.

Really, I want to overhaul all this anyway. I think when we take a pass over the regexp subsystem, a lot of this will get reworked.